File: OpenMarket/Models/Offer.swift
import Foundation

enum OfferStatus: String, Codable {
    case pending, accepted, declined, countered, withdrawn
}

struct Offer: Codable, Identifiable {
    let id: Int
    let listingID: Int
    let buyerID: Int
    let sellerID: Int
    let amount: Double
    let note: String
    let status: OfferStatus
    let counterAmount: Double?
    let buyerName: String?
    let listingTitle: String?
    let sellerName: String?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, amount, note, status
        case listingID     = "listing_id"
        case buyerID       = "buyer_id"
        case sellerID      = "seller_id"
        case counterAmount = "counter_amount"
        case buyerName     = "buyer_name"
        case listingTitle  = "listing_title"
        case sellerName    = "seller_name"
        case createdAt     = "created_at"
        case updatedAt     = "updated_at"
    }
}

struct CreateOfferRequest: Encodable {
    let listingID: Int
    let amount: Double
    let note: String

    enum CodingKeys: String, CodingKey {
        case listingID = "listing_id"
        case amount, note
    }
}

struct RespondToOfferRequest: Encodable {
    let action: String
    let counterAmount: Double?

    enum CodingKeys: String, CodingKey {
        case action
        case counterAmount = "counter_amount"
    }
}

File: backend/models/offer.go
package models

import "time"

type OfferStatus string

const (
	OfferStatusPending   OfferStatus = "pending"
	OfferStatusAccepted  OfferStatus = "accepted"
	OfferStatusDeclined  OfferStatus = "declined"
	OfferStatusCountered OfferStatus = "countered"
	OfferStatusWithdrawn OfferStatus = "withdrawn"
)

type Offer struct {
	ID            uint        `json:"id"`
	ListingID     uint        `gorm:"index;not null" json:"listing_id"`
	BuyerID       uint        `gorm:"index;not null" json:"buyer_id"`
	SellerID      uint        `gorm:"index;not null" json:"seller_id"`
	Amount        float64     `gorm:"not null" json:"amount"`
	Note          string      `json:"note"`
	Status        OfferStatus `gorm:"default:pending;not null" json:"status"`
	CounterAmount *float64    `json:"counter_amount,omitempty"`
	BuyerName     string      `gorm:"-" json:"buyer_name,omitempty"`
	CreatedAt     time.Time   `json:"created_at"`
	UpdatedAt     time.Time   `json:"updated_at"`
}
